crates/proto/src/rr/domain/name.rs

```diff
@@ -609,8 +609,19 @@ impl Name {
 
         if canonical {
             for label in labels {
+                labels_written.push(encoder.offset());
                 encoder.emit_character_data(label)?;
             }
+            let last_index = encoder.offset();
+            // now search for other labels already stored matching from the beginning label, strip then to the end
+            //   if it's not found, then store this as a new label
+            for label_idx in &labels_written {
```

Comment:
I think the code changes look good, but I wonder if we could reorganize it a little bit to ensure that labels are always being added in the same way both here, and down below. Would it need a new method extracted? I see that the logo is a little different, but it would be nice to try and not duplicate it.

Reply:
Missed this in my emails sorry, I have updated the code to refactor and remove some of the repeated lines. I decided to write it inline because after the refactor I thought it looked a little cleaner. What do you think?

Reply:
oh, interesting... I think this is a good change.